### PR TITLE
Optimize tracer assembly

### PR DESCRIPTION
**Pull request by akva2**
This is a continuation of https://github.com/OPM/opm-simulators/pull/4147

As I could not find a model with multiple phase tracers, I butchered SPE10-MOD01-01
```diff
--- SPE10-MOD01-01.DATA 2022-10-05 15:07:11.130820964 +0200
+++ SPE10-MOD01-01-TRACERS.DATA 2022-10-07 11:17:36.771923832 +0200
@@ -120,6 +120,10 @@
 --
 --       SWITCH ON THE UNIFIED INPUT FILES OPTION
 --
+
+TRACERS
+  2 2 2 /
+
 UNIFIN                                                                          
 --
 --       SWITCH ON THE UNIFIED OUTPUT FILES OPTION
@@ -347,6 +351,12 @@
 ROCK                                                                                    
          6000      1.0E-06                                 / TAKEN FROM MODEL 2
 
+TRACER
+ WT1  WAT 'g' /
+ WT2  OIL 'g' /
+ WT3  GAS 'g' /
+/
+
 -- =================================================================================================================================
 -- 
 -- SOLUTION SECTION 
@@ -364,6 +374,21 @@
 RPTRST                                                                           
          BASIC=5                                                               /
 
+-- Initial tracer concentration vs depth for tracer WT1
+TVDPFWT1
+ 1000  0.0
+ 2500  0.0 /
+
+-- Initial tracer concentration vs depth for tracer WT2
+TVDPFWT2
+ 1000  0.0
+ 2500  0.0 /
+
+-- Initial tracer concentration vs depth for tracer WT3
+TVDPFWT3
+ 1000  0.0
+ 2500  0.0 /
+
 -- =================================================================================================================================
 -- 
 -- SUMMARY SECTION 
```

And ran some benchmarks. The p-value given is compared to the previous column.
|procs| 0a3ede140  |      c3ddf59e7      |      08b38b155      |      c274623be      |
|----:|------------|---------------------|---------------------|---------------------|
|    1|6.31 +- 0.06|4.43 +- 0.05 (p=0.00)|4.32 +- 0.04 (p=0.00)|2.85 +- 0.01 (p=0.00)|
|    4|3.04 +- 0.02|2.45 +- 0.03 (p=0.00)|2.42 +- 0.02 (p=0.03)|1.94 +- 0.01 (p=0.00)|
|    8|3.63 +- 0.03|3.25 +- 0.01 (p=0.00)|3.24 +- 0.02 (p=0.64)|2.93 +- 0.02 (p=0.00)|

### CODE REVIEW

**Comment by bska** [comment]
> This is a continuation of #4147

Does that mean that this PR includes the changes of #4147?

**Comment by bska** [comment]
>> > This is a continuation of #4147
>> 
>> Does that mean that this PR includes the changes of #4147?
>
> yes, first 4 commits are from the #4147

Very good.  Maybe we can put this PR in a "draft" state until the upstream one is merged, then?

**Comment by bska** [comment]
jenkins build this please

**Comment by bska** [comment]
benchmark please

**Comment by ytelses** [comment]
Benchmark result overview: 




| Test | Configuration | Relative |
| ---- | ------------- | -------- |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=1823

**Comment by bska** [comment]
@akva2: Is there more work incoming here or is this PR ready for review now?

**Comment by bska** [comment]
jenkins build this please

File: ebos/ecltracermodel.hh [hunk 1]
```diff
@@ -154,11 +155,20 @@ public:
                 gas_.addTracer(tracerIdx, this->tracerConcentration_[tracerIdx]);
             }
         }
+        TracerMatrix* base = this->tracerMatrix_.get();
```
Comment (bska) [inline ebos/ecltracermodel.hh:158]: Could you maybe add a comment here to remind the reader that `base` is guaranteed to remain valid even when moving from `tracerMatrix_`?

**Review by bska — COMMENTED** [review]
Overall, this looks like a good improvement, and your performance numbers are very encouraging.  There are still some things that we might improve here&mdash;e.g, a lot more objects could be `const` and the `wellTracerRate_` is a very cumbersome data structure, but those are not directly related to this work.

If you add a comment concerning the validity of the raw pointer and ownership I'll be happy to merge this.

File: ebos/ecltracermodel.hh (hunk 1 above)
Comment (akva2) [reply ebos/ecltracermodel.hh:160]: done.

**Comment by akva2** [comment]
jenkins build this please

**Comment by akva2** [comment]
jenkins build this please

**Review by bska — APPROVED** [review]
Thank you for the updates.  I'll merge this into the master branch now.

**Pull request merged**